sk_sidekick/lib/src/commands/bump_version_command.dart

```diff
@@ -0,0 +1,200 @@
+
+    if (bumpMinor && bumpPatch) {
+      error('Either bump minor or patch');
+    }
+    if (!bumpMinor && !bumpPatch) {
+      // default to minor bump
+      bumpMinor = true;
+    }
+
+    final packagePath = argResults!.rest.firstOrNull ?? Directory.current.path;
+    final package = DartPackage.fromDirectory(Directory(packagePath));
+    if (package == null) {
+      throw 'Could not find a package in $packagePath';
+    }
+
+    final pubspecFile = package.pubspec;
+    final pubSpec = PubSpec.fromFile(pubspecFile.path);
+    final version = pubSpec.version!;
+
+    final newVersion = () {
+      if (bumpMajor) {
+        return version.nextMajor;
+      }
+      if (bumpMinor) {
+        return version.nextMinor;
+      }
+      if (bumpPatch) {
+        return version.nextPatch;
+      }
+
+      // default
+      return version.nextMajor;
+    }();
+
+    void applyModifications() {
+      for (final modification in _modifications) {
+        modification.call(package, version, newVersion);
+      }
+    }
+
+    bool bumped = false;
+    if (commit) {
+      final detachedHEAD = 'git symbolic-ref -q HEAD'
+          .start(progress: Progress.printStdErr(), nothrow: true);
+      if (detachedHEAD.exitCode != 0) {
+        printerr(
+          'You are in "detached HEAD" state. '
+          'Not committing version bump',
+        );
+      } else {
+        commitFileModifications(
+          applyModifications,
+          commitMessage: 'Bump version to $newVersion',
+        );
+        bumped = true;
+      }
+    }
+    if (!bumped) {
+      applyModifications();
+    }
+    print(green('Bumped ${package.name} from $version to $newVersion'));
+  }
+
+  /// Updates the version in pubspec.yaml
+  void bumpPubspecVersion(
+    DartPackage package,
+    Version oldVersion,
+    Version newVersion,
+  ) {
+    // Update version in pubspec.yaml
+    package.pubspec.replaceSectionWith(
+      startTag: 'version:',
+      endTag: '\n',
+      content: ' $newVersion',
+    );
+  }
+}
+
+typedef FileModification = void Function(
+  DartPackage package,
+  Version oldVersion,
+  Version newVersion,
+);
+
+/// Commits only the file changes that have been done in [block]
+void commitFileModifications(
+  void Function() block, {
+  required String commitMessage,
+}) {
+  // stash changes
+  'git stash save --include-untracked'.start(progress: Progress.printStdErr());
+
+  try {
+    // apply modifications
+    block();
+
+    // commit
+    'git add -A'.start(progress: Progress.printStdErr());
+    'git commit -m "$commitMessage" --no-verify'
+        .start(progress: Progress.printStdErr());
+    'git --no-pager log -n1 --oneline'.run;
+  } catch (e) {
+    printerr('Detected error, discarding modifications');
+    // discard all modifications
+    'git reset --hard'.start(progress: Progress.printStdErr());
+    rethrow;
+  } finally {
+    // restore changes
+    'git stash pop 0'.start(progress: Progress.printStdErr());
```

Comment:
This threw when I did the following:

- check out `sk-cli`  (18798dad5501fbfd5f6aff75812b946bfadc9b71)
- run `./sk bump-version sidekick_core --commit` in `/Users/pepe/repos/sidekick`

```
❯ ./sk bump-version sidekick_core --commit
2593690 (HEAD -> sk-cli) Bump version to 0.12.0
Unhandled exception:
git stash pop 0 
exit: 1
reason: The command [git] with args [stash, pop, 0] failed with exitCode: 1 workingDirectory: /Users/pepe/repos/sidekick
#0      waitForEx (package:dcli/src/util/wait_for_ex.dart:67)
#1      RunnableProcess.processUntilExit (package:dcli/src/util/runnable_process.dart:460)
#2      RunnableProcess.run (package:dcli/src/util/runnable_process.dart:180)
#3      start (package:dcli/src/functions/run.dart:249)
#4      StringAsProcess.start (package:dcli/src/util/string_as_process.dart:196)
#5      commitFileModifications (package:sk_sidekick/src/commands/bump_version_command.dart:159)
#6      BumpVersionCommand.run (package:sk_sidekick/src/commands/bump_version_command.dart:101)
```

Reply:
Yes, I've seen that before. It happens when you have local changes in the `pubspec.yaml`, then applying the patch doesn't work.
I'll check if I can somehow merge the patch.

Reply:
Either that or a clearer error message would be nice :)

Reply:
I think I know what the problem is.
The error is thrown even when I don't have any local changes in the pubspec.yaml.
When there are no local changes and you run `git stash save --include-untracked`, no stash is created because there are no local changes to save.
However, `git stash pop 0` is always called, even when no stash was created previously. This either causes a completely unrelated stash to be applied and deleted or throws an error when there are no stashes at all.

We could fix this by running `git stash | wc -l` before and after running `git stash save --include-untracked` to check whether a new stash has been added. Only if a new stash has been added, pop the last stash with `git stash pop 0`
 
```
❯ git status
On branch sk-cli
Your branch is up to date with 'origin/sk-cli'.

nothing to commit, working tree clean
❯ git rev-parse HEAD
9937e6d2c1e99b80353f606c2fda4a91aac46f1b
❯ ./sk bump-version sidekick_core --commit
9cead89 (HEAD -> sk-cli) Bump version to 0.12.0
Unhandled exception:
git stash pop 0 
exit: 1
reason: The command [git] with args [stash, pop, 0] failed with exitCode: 1 workingDirectory: /Users/pepe/repos/sidekick
#0      waitForEx (package:dcli/src/util/wait_for_ex.dart:67)
#1      RunnableProcess.processUntilExit (package:dcli/src/util/runnable_process.dart:460)
#2      RunnableProcess.run (package:dcli/src/util/runnable_process.dart:180)
#3      start (package:dcli/src/functions/run.dart:249)
#4      StringAsProcess.start (package:dcli/src/util/string_as_process.dart:196)
#5      commitFileModifications (package:sk_sidekick/src/commands/bump_version_command.dart:156)
#6      BumpVersionCommand.run (package:sk_sidekick/src/commands/bump_version_command.dart:98)
#7      CommandRunner.runCommand (package:args/command_runner.dart:209)
#8      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:119)
#9      new Future.sync (dart:async/future.dart:218)
#10     CommandRunner.run (package:args/command_runner.dart:119)
#11     SidekickCommandRunner.run (package:sidekick_core/sidekick_core.dart:129)
#12     runSk (package:sk_sidekick/sk_sidekick.dart:32)
#13     main (file:///Users/pepe/repos/sidekick/sk_sidekick/bin/main.dart:4)
#14     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:281)
#15     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184)
❯ git status
On branch sk-cli
Your branch is ahead of 'origin/sk-cli' by 1 commit.
  (use "git push" to publish your local commits)

Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add/rm <file>..." as appropriate to mark resolution)
        both modified:   sidekick_core/lib/sidekick_core.dart
        deleted by us:   sidekick_core/lib/src/sidekick_version_checker.dart
        deleted by us:   sidekick_core/lib/src/update_sidekick_cli.dart
        deleted by us:   sidekick_core/test/sidekick_version_checker_test.dart

no changes added to commit (use "git add" and/or "git commit -a")
```

Reply:
That's it! Kink of embarrassing tbh. In another similar [implementation](https://github.com/passsy/reformat-pre-commit-hook/blob/915f08e01b772013fe9ab4a9ea7e7d37a0a81ae3/pre-commit#L29) I did catch this error. Now I remember why I had to go the extra mile there.